plugins/rbac-backend/src/role-manager/member-list.ts

```diff
@@ -0,0 +1,124 @@
+  private roles: string[];
+
+  public constructor(name: string) {
+    this.name = name;
+    this.members = [];
+    this.roles = [];
+  }
+
+  /**
+   * addMembers will add members to the RoleMemberList
+   * @param members The members to be added.
+   */
+  public addMembers(members: string[]): void {
+    this.members = members;
+  }
+
+  /**
+   * addMember will add a single member to the RoleMemberList, skips adding the user in the
+   * event that they already exist in the members array.
+   * @param member The member to be added.
+   */
+  public addMember(member: string): void {
+    if (this.members.some(n => n === member)) {
+      return;
+    }
+    this.members.push(member);
+  }
+
+  /**
+   * hasMember will check if a particular member exists in the members array.
+   * @param name The member to be checked for.
+   */
+  public hasMember(name: string): boolean {
+    return this.members.includes(name);
+  }
+
+  /**
+   * deleteMember will remove a user from the members array.
+   * @param member The member to be removed.
+   */
+  public deleteMember(member: string): void {
+    this.members = this.members.filter(n => n !== member);
+  }
+
+  /**
+   * buildMembers will query the `casbin_rule` database table to ensure that the role
+   * that we have cached is up to date.
+   * This is important in multi node scenarios where the cached roles in role manager can become
+   * out of sync with the database.
+   * @param roleMemberList The RoleMemberList to be updated.
+   * @param client The database client.
+   */
+  public async buildMembers(
+    roleMemberList: RoleMemberList,
+    client: Knex,
+  ): Promise<void> {
+    try {
+      const members: string[] = await client
+        .table('casbin_rule')
+        .where('v1', this.name)
+        .pluck('v0');
+
+      roleMemberList.addMembers(members);
+    } catch (error) {
+      throw new Error(
+        `Unable to find members for the role ${this.name}. Cause: ${error}`,
+      );
+    }
+  }
+
+  /**
+   * getMembers will return the members of the RoleMemberList
+   * @returns The members.
+   */
+  getMembers(): string[] {
+    return this.members;
+  }
+
+  /**
+   * addRoles will add roles to the RoleMemberList
+   * @param roles The roles to be added.
+   */
+  public addRoles(roles: string[]): void {
+    this.roles = roles;
+  }
+
+  /**
+   * buildRoles will query the `casbin_rule` database table to quickly grab all of the
+   * roles that a particular user is attached to.
+   * @param roleMemberList The RoleMemberList to be updated.
+   * @param userAndGroups The user and groups to query with.
+   * @param client The database client.
+   */
+  public async buildRoles(
+    roleMemberList: RoleMemberList,
+    userAndGroups: string[],
+    client: Knex,
+  ): Promise<void> {
+    try {
+      const roles: string[] = await client
+        .table('casbin_rule')
+        .whereIn('v0', userAndGroups)
+        .pluck('v1');
+
+      roleMemberList.addRoles(roles);
```

Comment:
Sorry, but roles and members can be duplicated in the array. Are we going to handle that?

Reply:
I hope that our validator is good enough